### PR TITLE
Article: Measuring memory and CPU usage on Yocto & Armbian

### PR DESCRIPTION
**Pull request by pkonkol**

### CODE REVIEW

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+#### Comparison table
+
+|                        | load  | Mem usage    |
+|:-----------------------|:-----:|:------------:|
+|**before tests**        | 0.03  | 31MB / 244MB |
+|**during tests (max)**  | 0.12  | 35MB / 244MB |
+|**after tests**         | 0.04  | 31MB / 244MB |
+
+
+## Summary
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:205]: @pkonkol You could write something more about what is the origin of these differences between Yocto and Armbian.

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
+---
+title: Measuring memory and CPU usage on Yocto & Armbian
+abstract: 'Abstract first sentence.
+          Abstract second sentence.
+          Abstract third sentence.'
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:5]: @pkonkol abstract needed

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  - yocto
+  - armbian
+  - ram
+categories:
+  - IoT
+  - Miscellaneous
+
+---
+
+## Intro
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:25]: @pkonkol please add empty line between header and paragraph

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+This article will show you how to perform and interpret measurements of some of
+embedded systems computational resources. Case covered here will be an attemt
+to resolve a question whether our configuration will be enough for the given
+tasks, or an upgrade of the hardware specification will be necessary. As IoT and
+embedded systems often get very big in terms of device number, it is necessary
+to perfrom such analysis and match the specification exactly with the planned
+workload and the profit from such action will show as the system grows larger.
+
+Hardware configuration we will use is based on [RTE](https://3mdeb.com/products/open-source-hardware/rte/)
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:34]: @pkonkol based? Or it is exactly RTE?

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+with Orange Pi Zero 256MB. This is the default configuration with which it
+is shipped. Question may be made, whether 256MB of memory is sufficient, or
+should alternative, 512MB Orange Pi Zero be worth upgrading to. We will compare
+usage of memory on two operating systems - Armbian and Yocto. Memory will be
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:38]: @pkonkol link to Yocto meta layer and binary used would be useful, as well as the precise version and link to binary of Armbian

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+## Tools we will use
+
+### Basic tools
+
+These should be available on virtually every system.
+
+#### free
+`free` displays the total amount of free and used physical and swap
+memory in the system, as well as the buffers and caches used by  the
+kernel.  The  information  is  gathered by parsing /proc/meminfo. Running
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:50]:
```suggestion
kernel.  The information is gathered by parsing `/proc/meminfo`. Running
```

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+`armbianmonitor` is a simple CLI monitoring program for armbian system which
+shows CPU load, percentage usage of: cpu, sys (processes executing in kernel
+mode), usr (normal processes executing in user mode), nice (niced processes
+executing in user mode), io (waiting for I/O to complete), irq (servicing
+interrupts) and CPU temperature.
+
+`vmstat` reports information about processes, memory, paging, block IO, traps,
+disks and cpu activity. The first report produced gives averages since the last
+reboot. Additional reports give information on a sampling period of length
+delay. The process and memory reports are instantaneous in either case.
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:88]:
```suggestion
delay. The process and memory reports are instantaneous in either case.

```

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+KiB Swap:   131068 total,   131068 free,        0 used.   171404 avail Mem
+```
+
+#### Comparison table
+
+|                       |load   | CPU temp   | Mem usage    |
+|:----------------------|:-----:|:----------:|:------------:|
+|**before tests**       | 0.03  | 43.9&deg;C | 32MB / 242MB |
+|**during tests (max)** | 0.53  | 51&deg;C   | 65MB / 242MB |
+|**after tests**        | 0.12  | 44.2&deg;C | 33MB / 242MB |
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:149]: @pkonkol graphical diagram using Google Calc or OpenOffice would be appreciate and would say more to reader then plain numbers

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+KiB Swap:        0 total,        0 free,        0 used.   200580 avail Mem
+```
+
+#### Comparison table
+
+|                        | load  | Mem usage    |
+|:-----------------------|:-----:|:------------:|
+|**before tests**        | 0.03  | 31MB / 244MB |
+|**during tests (max)**  | 0.12  | 35MB / 244MB |
+|**after tests**         | 0.04  | 31MB / 244MB |
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:202]: @pkonkol why no temperature reading?

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+
+## Summary
+As we can see, 256MB of RAM is beyond enough for current feature-set on the
+RTE both on Armbian and Yocto, yet there are significant differences between
+them as the former had almost twice as high maximal memory usage as the latter.
+Similar difference was seen in the maximal spike in cpu load, which on Armbian
+was over 4 times that of Yocto.
+
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:211]: Redundant empty line

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 11]
```diff
@@ -0,0 +1,217 @@
+---
+title: Measuring memory and CPU usage on Yocto & Armbian
```
Comment (pietrushnic) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:2]: @pkonkol please consider to add infor about OrangePI in title

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md (hunk 4 above)
Comment (lwcislo) [reply blog/content/post/2020-04-26-measuring_memory_usage_rte.md:34]: Well, there is also the DUT, which is not relevant but required for regression.

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+  - ram
+categories:
+  - IoT
+  - Miscellaneous
+
+---
+
+## Intro
+This article will show you how to perform and interpret measurements of some of
+embedded systems computational resources. Case covered here will be an attemt
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:27]: @pkonkol attempt

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+to resolve a question whether our configuration will be enough for the given
+tasks, or an upgrade of the hardware specification will be necessary. As IoT and
+embedded systems often get very big in terms of device number, it is necessary
+to perfrom such analysis and match the specification exactly with the planned
+workload and the profit from such action will show as the system grows larger.
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:32]: @pkonkol I'd split this long sentence into two short ones

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+checked before and during the run of regression test suite.
+
+## Tools we will use
+
+### Basic tools
+
+These should be available on virtually every system.
+
+#### free
+`free` displays the total amount of free and used physical and swap
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:48]: @pkonkol add empty line after the header

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+it with `-m` option will print the values in mebibytes, which often get
+confused with megabytes. The diffrence is that the former are express the
+values as powers of two, while the latter express the values as powers of ten.
+
+To get real-time readings we will combine it with `watch` command, which will
+print the result of command supplied to it with the freqency specified by
+number passed to `-n` parameter in seconds. The final command is
+`watch -n 1 free -m`
+
+#### top
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:60]: @pkonkol missing empty line after the header

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+`top` program provides a dynamic real-time view of a running system. It can
+display system summary information as well as a list of tasks currently
+being managed by the Linux kernel. Its output is complete enough so that
+we won't be specifying any additional parameters.
+
+#### /sys/class/thermal/thermal_zone
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:66]: @pkonkol missing empty line after the header

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+Thremal sysfs provides us infromation about readings from various temperature
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:67]: Thermal

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+sernsors installed. Different sensors are available as thermal_zone[0-*] files.
+You can check sensor type by reading `type` file from thermal_zone directory.
+In our case the CPU thermal_zone has number `0`. Temperature is stored in
+`temp` file as an integer.  To get the output in a more accessible format
+of Celcius degrees we will use such command:
+```
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:73]: empty line before block of code

File: blog/data/authors/piotr.konkol.json [hunk 19]
```diff
@@ -0,0 +1,5 @@
+{
+    "name": "Piotr Konkol",
+    "bio": "Junior Test Automation Developer",
```
Comment (artur-rs) [inline blog/data/authors/piotr.konkol.json:3]: @pkonkol this bio is a little short (it is temporary right?). Also, `index.md` file is missing for a full profile support. Check the `README.md` for requirements.

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md (hunk 9 above)
Comment (pkonkol) [reply blog/content/post/2020-04-26-measuring_memory_usage_rte.md:203]: I've based my article on documentation not made by me, there was no temp reading there.

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md (hunk 5 above)
Comment (pkonkol) [reply blog/content/post/2020-04-26-measuring_memory_usage_rte.md:38]: @artur-rs Do you know if I could find such files somewhere?

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md (hunk 4 above)
Comment (pkonkol) [reply blog/content/post/2020-04-26-measuring_memory_usage_rte.md:34]: @pietrushnic I thought about this in a way that @lwcislo mentioned

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md (hunk 5 above)
Comment (artur-rs) [reply blog/content/post/2020-04-26-measuring_memory_usage_rte.md:39]: @pkonkol links sent in direct msg

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+necessary to perfrom such analysis and match the specification exactly with the
+planned workload and the profit from such action will show as the system grows
+larger.
+
+Hardware configuration we will use is based on [RTE](https://3mdeb.com/products/open-source-hardware/rte/)
+with Orange Pi Zero 256MB. This is the default configuration with which it
+is shipped. Question may be made, whether 256MB of memory is sufficient, or
+should alternative, 512MB Orange Pi Zero be worth upgrading to. We will compare
+usage of memory on two operating systems - Armbian and Yocto. We used Armbian
+version 5.32.170919. Armbian binaries can be downloaded [here](https://dl.armbian.com/_old/orangepizero/archive/)
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:40]: @pkonkol missing dot after `here` link

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+These should be available on virtually every system.
+
+#### free
+
+`free` displays the total amount of free and used physical and swap
+memory in the system, as well as the buffers and caches used by  the
+kernel.  The information is gathered by parsing `/proc/meminfo`. Running
+it with `-m` option will print the values in mebibytes, which often get
+confused with megabytes. The diffrence is that the former are express the
```
Comment (artur-rs) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:57]: @pkonkol difference

File: blog/content/authors/piotr-konkol/_index.md [hunk 22]
```diff
@@ -0,0 +1,11 @@
+---
+name: "Piotr Konkol"
+identifier: piotr.konkol
+photo: "/authors/piotr.konkol.png"
```
Comment (artur-rs) [inline blog/content/authors/piotr-konkol/_index.md:4]: @pkonkol .jpg

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+---
+
+## Intro
+
+This article will show you how to perform and interpret measurements of some of
+embedded systems computational resources. Case covered here will be an attempt
+to resolve a question whether our configuration will be enough for the given
+tasks, or an upgrade of the hardware specification will be necessary. IoT and
+embedded systems tend to often get very big in terms of device number. It is
+necessary to perfrom such analysis and match the specification exactly with the
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:31]: perform

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+planned workload and the profit from such action will show as the system grows
+larger.
+
+Hardware configuration we will use is based on [RTE](https://3mdeb.com/products/open-source-hardware/rte/)
+with Orange Pi Zero 256MB. This is the default configuration with which it
+is shipped. Question may be made, whether 256MB of memory is sufficient, or
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:37]: A question

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+should alternative, 512MB Orange Pi Zero be worth upgrading to. We will compare
+usage of memory on two operating systems - Armbian and Yocto. We used Armbian
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:39]: the usage

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+#### top
+
+`top` program provides a dynamic real-time view of a running system. It can
+display system summary information as well as a list of tasks currently
+being managed by the Linux kernel. Its output is complete enough so that
+we won't be specifying any additional parameters.
+
+#### /sys/class/thermal/thermal_zone
+
+Thermal sysfs provides us infromation about readings from various temperature
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:74]: information

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+sernsors installed. Different sensors are available as thermal_zone[0-*] files.
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:75]: sensors

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+You can check sensor type by reading `type` file from thermal_zone directory.
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:76]: the sensor type

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 29]
```diff
@@ -0,0 +1,222 @@
...
+
+![Yocto](/img/Yocto_load&mem_usage_chart.svg)
+
+## Summary
+As we can see, 256MB of RAM is beyond enough for current feature-set on the
+RTE both on Armbian and Yocto, yet there are significant differences between
+them as the former had almost twice as high maximal memory usage as the latter.
+Similar difference was seen in the maximal spike in cpu load, which on Armbian
+was over 4 times that of Yocto. As Yocto is not really an embedded Linux
+distribution, but a framework for creating your own, suited specifficaly to
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:211]: specifically

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+
+#### Comparison table
+
+![Yocto](/img/Yocto_load&mem_usage_chart.svg)
+
+## Summary
+As we can see, 256MB of RAM is beyond enough for current feature-set on the
+RTE both on Armbian and Yocto, yet there are significant differences between
+them as the former had almost twice as high maximal memory usage as the latter.
+Similar difference was seen in the maximal spike in cpu load, which on Armbian
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:209]: A similar

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 31]
```diff
@@ -0,0 +1,222 @@
...
+was over 4 times that of Yocto. As Yocto is not really an embedded Linux
+distribution, but a framework for creating your own, suited specifficaly to
+your goals and hardware it is able to provide much better performance. Armbian
+is a good choice for early prototyping, as it's popular and easy to use, but
+for final product it is worth to consider using Yocto.
```
Comment (lwcislo) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:214]: for the final product, it is worth considering using Yocto.

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 32]
```diff
@@ -0,0 +1,222 @@
...
+necessary to perform such analysis and match the specification exactly with the
+planned workload and the profit from such action will show as the system grows
+larger.
+
+Hardware configuration we will use is based on [RTE](https://3mdeb.com/products/open-source-hardware/rte/)
+with Orange Pi Zero 256MB. This is the default configuration with which it
+is shipped. A question may be made, whether 256MB of memory is sufficient, or
+should alternative, 512MB Orange Pi Zero be worth upgrading to. We will compare
+the usage of memory on two operating systems - Armbian and Yocto. We used Armbian
+version 5.32.170919. Armbian binaries can be downloaded [here](https://dl.armbian.com/_old/orangepizero/archive/).
```
Comment (KBanecka) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:40]: version:

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+[Yocto meta-rte](https://github.com/3mdeb/meta-rte) which we used can be found
+[here](https://cloud.3mdeb.com/index.php/s/myTkar9CgrgKG9m/download).
+Memory will be checked before and during the run of regression test suite.
+
+## Tools we will use
+
+### Basic tools
+
+These should be available on virtually every system.
```
Comment (KBanecka) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:49]: on virtually ?

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 34]
```diff
@@ -0,0 +1,222 @@
...
+To get real-time readings we will combine it with `watch` command, which will
+print the result of command supplied to it with the freqency specified by
+number passed to `-n` parameter in seconds. The final command is
+`watch -n 1 free -m`
+
+#### top
+
+`top` program provides a dynamic real-time view of a running system. It can
+display system summary information as well as a list of tasks currently
+being managed by the Linux kernel. Its output is complete enough so that
```
Comment (KBanecka) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:69]: completely

File: blog/content/post/2020-04-26-measuring_memory_usage_rte.md [hunk 35]
```diff
@@ -0,0 +1,222 @@
...
+we won't be specifying any additional parameters.
+
+#### /sys/class/thermal/thermal_zone
+
+Thermal sysfs provides us information about readings from various temperature
```
Comment (KBanecka) [inline blog/content/post/2020-04-26-measuring_memory_usage_rte.md:74]: provides us with

**Pull request merged**